pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,327 @@
+	}
+	if itops.BatchSize > int(g.m.MaxEntries()) {
+		itops.BatchSize = int(g.m.MaxEntries())
+	}
+
+	if BatchAPISupported() && !g.isPerCPU() && !itops.ForceSingleItem {
+		it := &genericMapBatchIterator[K, V]{
+			m:                            g.m,
+			batchSize:                    itops.BatchSize,
+			keys:                         make([]K, itops.BatchSize),
+			values:                       make([]V, itops.BatchSize),
+			valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+		}
+
+		// Do an initial copy of the keys/values slices to avoid allocations
+		it.keysCopy = it.keys
+		it.valuesCopy = it.values
+
+		return it
+	}
+
+	return &genericMapItemIterator[K, V]{
+		it:                           g.m.Iterate(),
+		valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+	}
+}
+
+type genericMapItemIterator[K interface{}, V interface{}] struct {
+	it                           *ebpf.MapIterator
+	valueTypeCanUseUnsafePointer bool
+}
+
+func (g *genericMapItemIterator[K, V]) Next(key *K, value *V) bool {
+	// we resort to unsafe.Pointers because by doing so the underlying eBPF
+	// library avoids marshaling the key/value variables while traversing the map
+	// However, in some cases (slices, structs) we need to pass the variable directly
+	// so that the library detects the type correctly
+	if g.valueTypeCanUseUnsafePointer {
+		return g.it.Next(unsafe.Pointer(key), unsafe.Pointer(value))
+	}
+
+	return g.it.Next(unsafe.Pointer(key), value)
+}
+
+func (g *genericMapItemIterator[K, V]) Err() error {
+	return g.it.Err()
+}
+
+type genericMapBatchIterator[K interface{}, V interface{}] struct {
+	m                            *ebpf.Map
+	batchSize                    int
+	cursor                       ebpf.BatchCursor
+	keys                         []K
+	values                       []V
+	keysCopy                     any // A pointer to keys of type "any", used to avoid allocations when calling BatchLookup
+	valuesCopy                   any
+	currentBatchSize             int
+	inBatchIndex                 int
+	err                          error
+	totalCount                   int
+	lastBatch                    bool
+	valueTypeCanUseUnsafePointer bool
+}
+
+func (g *genericMapBatchIterator[K, V]) Next(key *K, value *V) bool {
+	// Safety check to avoid an infinite loop
+	if g.totalCount >= int(g.m.MaxEntries()) {
+		return false
+	}
+
+	// We have finished all the values in the current batch (or there wasn't any batch
+	// to begin with, with g.currentBatchSize == 0), so we need to fetch the next batch
+	if g.inBatchIndex >= g.currentBatchSize {
+		if g.lastBatch {
+			return false
+		}
+
+		// Important! If we pass here g.keys/g.values, Go will create a copy of the slice instance
+		// and will generate extra allocations. I am not entirely sure why it is doing that.
+		g.currentBatchSize, g.err = g.m.BatchLookup(&g.cursor, g.keysCopy, g.valuesCopy, nil)
```

Comment:
I think it is fine to provide the origianl keys and values, the slice is a very slim struct itself just holding the pointer to the underlying buffer. the reason go allocates it when passed by value is due to the fact the slice is immutable in golang

moreover have you seen [this](https://github.com/cilium/ebpf/blob/main/map.go#L964-L965) comment in cilium/ebpf for the `BatchLookup` function?

Reply:
also I am a bit concerned about the [next lines](https://github.com/cilium/ebpf/blob/main/map.go#L970-972) in that comment

Reply:
Yes, the reason I did the copy first was mostly to have 0 allocations per iteration, Bryce told me to take care with that. It's true that it's not much, but I think this approach gives better performance and memory usage (only one copy is done instead of one per batch). Regarding the first comment, I think the tests cover all the cases, it's true that if you pass a pointer (e.g. `&g.keysCopy`) it complains about it.

For the second comment, it is indeed a little bit weird. For example, a batch might return less items than requested even if it's not the last one. It's also unintuitive that the return code for the last batch is ErrKeyNotExists, although in my tests I've seen that only happens always for hash maps, array maps don't always return ErrKeyNotExists for the last batch, it just returns 0 elements sometimes.

Reviewing both the code of the library and the usage in code, I think the tests cover the use cases we have in the agent (I just uploaded a test for the Array case).

Reply:
I was more referring to:

> // Warning: This API is not very safe to use as the kernel implementation for
> // batching relies on the user to be aware of subtle details with regarding to
> // different map type implementations.

rather than return code

Reply:
I assumed those details is what they were referring to, as that was the only difference I noticed while working with it and looking at the kernel code. However, we could maybe change the implementation so that the default is single-item iteration, so that batch iteration has to be selected explicitly and people using that can test that it works in their use case.

Reply:
It looks like you can do this and avoid the allocations:
```suggestion
		g.currentBatchSize, g.err = g.m.BatchLookup(&g.cursor, any(g.keys), any(g.values), nil)
```

Reply:
Changing that breaks [this test](https://github.com/DataDog/datadog-agent/blob/9bb26d7713c1a52cde6696c21199d82e2609fa67/pkg/ebpf/generic_map_test.go#L342) that checks that there are not extra allocations per iteration. We could skip that test until cilium/ebpf fixes this behavior in https://github.com/cilium/ebpf/issues/1290? I can also document this in the code, keep an eye on that issue and send a PR when they fix that.